libnmstate/nm/device.py

```diff
@@ -173,62 +173,41 @@ def delete_device(nmdev):
 
 def _safe_delete_device_async(nmdev):
     mainloop = nmclient.mainloop()
-    user_data = mainloop, nmdev
+    user_data = mainloop, nmdev, nmdev.get_iface()
     nmdev.delete_async(
         mainloop.cancellable, _delete_device_callback, user_data
     )
 
 
 def _delete_device_callback(src_object, result, user_data):
-    mainloop, nmdev = user_data
+    mainloop, nmdev, iface = user_data
+    error = None
     try:
-        success = src_object.delete_finish(result)
-    # pylint: disable=catching-non-exception
-    except nmclient.GLib.GError as e:
-        # pylint: enable=catching-non-exception
-        if e.matches(
-            nmclient.Gio.DBusError.quark(),
-            nmclient.Gio.DBusError.UNKNOWN_METHOD,
-        ) or (
-            e.matches(
-                nmclient.NM.DeviceError.quark(),
-                nmclient.NM.DeviceError.NOTSOFTWARE,
-            )
-            and nmdev.is_software()
-        ):
-            logging.debug(
-                "Device %s has been already deleted: error=%s",
-                nmdev.get_iface(),
-                e,
-            )
-            mainloop.execute_next_action()
-        else:
-            mainloop.quit(
-                "Device deletion failed on {}: error={}".format(
-                    nmdev.get_iface(), e
-                )
-            )
-        return
+        src_object.delete_finish(result)
     except Exception as e:
-        if mainloop.is_action_canceled(e):
+        error = e
+        if mainloop.is_action_canceled(error):
             logging.debug(
-                "Device deletion aborted on %s: error=%s", nmdev.get_iface(), e
+                "Device deletion aborted on %s: error=%s", iface, error
             )
-        else:
-            mainloop.quit(
-                "Device deletion failed on {}: error={}".format(
-                    nmdev.get_iface(), e
-                )
+            return
+
+    mainloop.iterate_all()
```

Comment:
This requires an explanation, probably in the docttext of that function.

Reply:
IMHO the commit message fits better because it explains the other code change as well.

Reply:
I added it to the commit message.

Reply:
I removed this for now, maybe it is not needed. My debugging session was so long ago, I do not remember all the details.